hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/DispatchingEvmFrameState.java

```diff
@@ -87,72 +87,90 @@ public class DispatchingEvmFrameState implements EvmFrameState {
-            @NonNull final WritableKVState<EntityNumber, Bytecode> bytecode) {
-        this.storage = requireNonNull(storage);
-        this.bytecode = requireNonNull(bytecode);
-        this.dispatch = requireNonNull(dispatch);
+            @NonNull final HederaNativeOperations nativeOperations,
+            @NonNull final ContractStateStore contractStateStore) {
+        this.nativeOperations = requireNonNull(nativeOperations);
+        this.contractStateStore = requireNonNull(contractStateStore);
     }
 
     @Override
     public void setStorageValue(final long number, @NonNull final UInt256 key, @NonNull final UInt256 value) {
         final var slotKey = new SlotKey(number, tuweniToPbjBytes(requireNonNull(key)));
-        final var oldSlotValue = storage.get(slotKey);
-        // Ensure we don't change any prev/next keys until committing the final WorldUpdater
+        final var oldSlotValue = contractStateStore.getSlotValue(slotKey);
+        // Ensure we don't change any prev/next keys until the base commit
         final var slotValue = new SlotValue(
                 tuweniToPbjBytes(requireNonNull(value)),
                 oldSlotValue == null ? com.hedera.pbj.runtime.io.buffer.Bytes.EMPTY : oldSlotValue.previousKey(),
                 oldSlotValue == null ? com.hedera.pbj.runtime.io.buffer.Bytes.EMPTY : oldSlotValue.nextKey());
-        storage.put(slotKey, slotValue);
+        // We don't call remove() here when the new value is zero, again because we
+        // want to preserve the prev/next key information until the base commit; only
+        // then will we remove the zeroed out slot from the K/V state
+        contractStateStore.putSlot(slotKey, slotValue);
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public @NonNull UInt256 getStorageValue(final long number, @NonNull final UInt256 key) {
         final var slotKey = new SlotKey(number, tuweniToPbjBytes(requireNonNull(key)));
-        return valueOrZero(storage.get(slotKey));
+        return valueOrZero(contractStateStore.getSlotValue(slotKey));
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public @NonNull UInt256 getOriginalStorageValue(final long number, @NonNull final UInt256 key) {
         final var slotKey = new SlotKey(number, tuweniToPbjBytes(requireNonNull(key)));
-        return valueOrZero(storage.getOriginalValue(slotKey));
+        return valueOrZero(contractStateStore.getOriginalSlotValue(slotKey));
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
     public @NonNull List<StorageAccesses> getStorageChanges() {
         final Map<Long, List<StorageAccess>> modifications = new TreeMap<>();
-        storage.modifiedKeys().forEach(slotKey -> modifications
+        contractStateStore.getModifiedSlotKeys().forEach(slotKey -> modifications
                 .computeIfAbsent(slotKey.contractNumber(), k -> new ArrayList<>())
                 .add(StorageAccess.newWrite(
```

Comment:
The two transforms pbj<->besu(tuweni) could be pulled out into a matched set of methods.